### DISCUSSION TITLE
UERANSIM Feature Question

### DISCUSSION THREAD

**Discussion by jpsla94** [aligungr/UERANSIM#304] [Q&A]
Hello All,

Does UERANSIM have a feature to simulate device traffic? If not, does it work with any specific traffic simulation software?

Thanks in advance.

**Comment by aligungr** [comment]
Hello

UERANSIM does not have builtin feature for traffic generation. But some external traffic generators should be able to utilize the `uesimtunX` interface.

Regards

**Reply by jpsla94** [reply]
Thanks for reply, @aligungr. Is it possible to share some documentation about this interface so I can see what I am working with?

**Reply by aligungr** [reply]
Actually there is no special documentation for `uesimtunX` interface. Because it's just a standard Linux network adapter like any other network adapters in your system such as `enp0s3`, `wlan0` etc.

I didn't use a traffic generator in the past, but my guess is that these tools should be able to select an interface such as `enp0s3`, `wlan0`, or in our case `uesimtunX`

**Reply by jpsla94** [reply]
Ok, I didn´t understand at first that you were referring to networks interfaces 😕. None the less, thanks for your help, @aligungr. I will look into this.

**Reply by aligungr** [reply]
Sure, I could not explain it, sorry :)

**Reply by pstavirs** [reply]
You can use [Ostinato](https://ostinato.org) to send traffic on the `tunX` interface created by UERANSIM.

Note that Ostinato _by default_ generates ethernet packets while tunX is an "ip" interface. You can generate pure IP traffic from Ostinato as shown below

![ueransim-raw-ip](https://github.com/aligungr/UERANSIM/assets/8395079/3613a63c-30de-483b-87c9-1b4e5228f941)

Full disclosure: I'm the creator of Ostinato
